rfd/0004-workspace-and-prebuild-storage.md

```diff
@@ -0,0 +1,97 @@
+---
+authors: Hugo Dutka (contact@hugodutka.com)
+state: draft
+---
+
+# RFD 4 - Workspace and prebuild storage management
+
+## What
+
+Specify high-level workspace and prebuild storage management in Hocus using [the local block registry](./0001-block-registry.md) and a remote OCI registry.
+
+## Why
+
+Goals:
+
+1. Store data without duplication between workspaces that use the same prebuilds.
+2. Allow sharing prebuilds between users on different machines to avoid redundant builds.
+3. Enable fast workspace start-up, even without local caching.
+
+Storage backed by OCI images and powered by [overlaybd](https://github.com/containerd/overlaybd) with lazy pulling fits these requirements.
+
+## Details
+
+### Why OCI images?
+
+Using OCI images lets us benefit from the OCI ecosystem. In particular, we make heavy use of overlaybd, which underpins our local block registry. Overlaybd supports lazy pulling, which allows us to run a workspace without fully downloading its backing files, and then download the rest in the background. We can also use standardized OCI registries for long-term storage and transferring images between machines, eliminating the need for a custom solution.
+
+### Types of images Hocus needs
+
+Hocus needs to keep track of the following types of images:
+
+- git repository images
+- prebuild images, two types:
+  - code images (based on git repository images)
+  - os images
+- workspace images, two types:
+  - code images (based on prebuild code images)
+  - os images (based on prebuild os images)
+
+In total, there are 5 types of images Hocus manages.
+
+### Local and remote storage of images
+
+Prebuild images are pushed to a user-configurable OCI registry after they are built. Every prebuild has a tag that is partly deterministic, which allows us to check if it's been built before and if the agent can download it instead of building. After they are downloaded or built, prebuild images are stored on disk until the associated prebuild is archived. Workspace images are based on prebuild images, but for now they are only stored locally and never pushed to a remote location. They are deleted once a workspace is deleted. Git repository images are not pushed to a remote registry for now.
```

Comment:
+1 What does it mean to archive a prebuild? Deleting it from the remote oci registry or locally?